src/Workspaces/Core/Portable/Rename/ConflictEngine/ConflictResolver.SingleSymbolRenameSession.cs

```diff
@@ -0,0 +1,137 @@
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeCleanup;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Rename.ConflictEngine
+{
+    internal static partial class ConflictResolver
+    {
+        private class SingleSymbolRenameSession : Session
+        {
+            private readonly string _originalText;
+            private readonly string _replacementText;
+
+            public static async Task<SingleSymbolRenameSession> CreateAsync(
+                SymbolicRenameLocations renameLocationSet,
+                Location renameSymbolDeclarationLocation,
+                string replacementText,
+                ImmutableArray<SymbolKey> nonConflictSymbolKeys,
+                CodeCleanupOptionsProvider fallbackOptions,
+                CancellationToken cancellationToken)
+            {
+                var originalText = renameLocationSet.Symbol.Name;
+                var renameLocations = renameLocationSet.Locations;
+                var symbol = renameLocationSet.Symbol;
+
+                // only process documents which possibly contain the identifiers.
+                var documentIdOfRenameSymbolDeclaration = renameLocationSet.Solution.GetRequiredDocument(renameSymbolDeclarationLocation.SourceTree!).Id;
+
+                using var _ = PooledDictionary<DocumentId, DocumentRenameInfo>.GetInstance(out var documentIdToRenameInfoBuilder);
+                var (documentsIdsToBeCheckedForConflict, possibleNameConflicts) = await FindDocumentsAndPossibleNameConflictsAsync(
+                    renameLocationSet,
+                    replacementText,
+                    originalText,
+                    cancellationToken).ConfigureAwait(false);
+
+                var replacementTextValid = IsIdentifierValid_Worker(
+                    renameLocationSet.Solution,
+                    replacementText,
+                    documentsIdsToBeCheckedForConflict.Select(documentId => documentId.ProjectId).Distinct());
+
+                var documentIdToRenameLocations = renameLocations
+                    .GroupBy(location => location.DocumentId)
+                    .ToDictionary(grouping => grouping.Key);
+                foreach (var documentId in documentsIdsToBeCheckedForConflict)
+                {
+                    using var documentRenameInfoBuilder = new DocumentRenameInfo.Builder();
+
+                    // Conflict checking documents is a superset of the rename locations. In case this document is not a documents of rename locations,
+                    // just passing an empty rename information to check for conflicts.
+                    var renameLocationsInDocument = documentIdToRenameLocations.ContainsKey(documentId)
```

Comment:
This is a little hard to follow. Reading below, it looks like the only work left is 

```c#
                    documentRenameInfoBuilder.AddRenamedSymbol(symbol, replacementText, replacementTextValid, possibleNameConflicts);
                    documentIdToRenameInfoBuilder[documentId] = documentRenameInfoBuilder.ToRenameInfo();
```

Everything else is skipped because of the empty array. Can we restructure this to clarify more? Or move to another method

Reply:
Yup this is a pretty tricky part of the conflict resolution. (I was thinking like this at the beginning)
Basically `ConflictCheckingDocument` = `All the renaming documents` + `All the possible naming conflict documents`, the second part contains some documents which could just have the same text in some of the identifiers. And we need to visit them to annotate the potential conflict locations.

Let me refactor this part a little bit to make it more clear